File: ProjectTrackervb/ProjectTracker.Library/Assignment.vb
Imports Csla.Validation
Imports System.Reflection

Friend Interface IHoldRoles
  Property Role() As Integer
End Interface

Friend Module Assignment

#Region " Business Methods "

  Public Function GetDefaultAssignedDate() As Date

    Return Today

  End Function

#End Region

#Region " Validation Rules "

  ''' <summary>
  ''' Ensure the Role property value exists
  ''' in RoleList
  ''' </summary>
  Public Function ValidRole( _
    ByVal target As Object, ByVal e As RuleArgs) As Boolean

    Dim role As Integer = CType(target, IHoldRoles).Role

    If RoleList.GetList.ContainsKey(role) Then
      Return True

    Else
      e.Description = "Role must be in RoleList"
      Return False
    End If

  End Function

#End Region

#Region " Data Access "

  Public Function AddAssignment( _
    ByVal projectId As Guid, _
    ByVal resourceId As Integer, ByVal assigned As SmartDate, _
    ByVal role As Integer) As Byte()

    Using ctx = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager(ProjectTracker.DalLinq.Database.PTracker, True)
      Dim lastChanged As System.Data.Linq.Binary = Nothing
      ctx.DataContext.addAssignment( _
        projectId, resourceId, assigned, role, lastChanged)
      Return lastChanged.ToArray
    End Using

  End Function

  Public Function UpdateAssignment( _
    ByVal projectId As Guid, ByVal resourceId As Integer, _
    ByVal assigned As SmartDate, ByVal newRole As Integer, _
    ByVal timestamp() As Byte) As Byte()

    Using ctx = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager(ProjectTracker.DalLinq.Database.PTracker, True)
      Dim lastChanged As System.Data.Linq.Binary = Nothing
      ctx.DataContext.UpdateAssignment( _
        projectId, resourceId, assigned, newRole, timestamp, lastChanged)
      Return lastChanged.ToArray
    End Using

  End Function

  Public Sub RemoveAssignment( _
    ByVal projectId As Guid, _
    ByVal resourceId As Integer)

    Using ctx = ContextManager(Of ProjectTracker.DalLinq.PTrackerDataContext).GetManager(ProjectTracker.DalLinq.Database.PTracker, True)
      Dim lastChanged As System.Data.Linq.Binary = Nothing
      ctx.DataContext.DeleteAssignment( _
        projectId, resourceId)
    End Using

  End Sub

#End Region

End Module

File: Samples/CslaNet/cs/ProjectTrackercs/ProjectTracker.Library/Assignment.cs
using System;
using Csla;
using Csla.Data;
using Csla.Rules;

namespace ProjectTracker.Library
{
  internal interface IHoldRoles
  {
    int Role { get; set; }
  }

  internal static class Assignment
  {
    #region  Business Methods

    public static System.DateTime GetDefaultAssignedDate()
    {
      return System.DateTime.Today;
    }

    #endregion

    #region  Validation Rules

    /// <summary>
    /// Ensure the Role property value exists
    /// in RoleList
    /// </summary>
    public class ValidRole : BusinessRule
    {
      protected override void Execute(RuleContext context)
      {
        var target = (IHoldRoles)context.Target;
        int role = target.Role;

        if (!RoleList.GetList().ContainsKey(role))
          context.AddErrorResult("Role must be in RoleList");
      }
    }

    #endregion

    #region  Data Access

    public static byte[] AddAssignment(
      Guid projectId, 
      int resourceId, 
      SmartDate assigned, 
      int role)
    {
      using (var ctx = 
        ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.
        GetManager(ProjectTracker.DalLinq.Database.PTracker))
      {
        System.Data.Linq.Binary lastChanged = null;
        ctx.DataContext.addAssignment(
          projectId, 
          resourceId, 
          assigned, 
          role, 
          ref lastChanged);
        return lastChanged.ToArray();
      }
    }

    public static byte[] UpdateAssignment(Guid projectId, int resourceId, SmartDate assigned, int newRole, byte[] timestamp)
    {
      using (var ctx = ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.GetManager(ProjectTracker.DalLinq.Database.PTracker))
      {
        System.Data.Linq.Binary lastChanged = null;
        ctx.DataContext.updateAssignment(projectId, resourceId, assigned, newRole, timestamp, ref lastChanged);
        return lastChanged.ToArray();
      }
    }

    public static void RemoveAssignment(Guid projectId, int resourceId)
    {
      using (var ctx = ContextManager<ProjectTracker.DalLinq.PTrackerDataContext>.GetManager(ProjectTracker.DalLinq.Database.PTracker))
      {
        ctx.DataContext.deleteAssignment(projectId, resourceId);
      }
    }

    #endregion

  }
}